api/renewer.go

```diff
@@ -0,0 +1,303 @@
+	}
+
+	grace := i.Grace
+	if grace == 0 {
+		grace = DefaultRenewerGrace
+	}
+
+	random := i.Rand
+	if random == nil {
+		random = rand.New(rand.NewSource(int64(time.Now().Nanosecond())))
+	}
+
+	renewBuffer := i.RenewBuffer
+	if renewBuffer == 0 {
+		renewBuffer = DefaultRenewerRenewBuffer
+	}
+
+	return &Renewer{
+		client:  c,
+		secret:  secret,
+		grace:   grace,
+		random:  random,
+		doneCh:  make(chan error, 1),
+		renewCh: make(chan *RenewOutput, renewBuffer),
+
+		stopped: false,
+		stopCh:  make(chan struct{}),
+	}, nil
+}
+
+// DoneCh returns the channel where the renewer will publish when renewal stops.
+// If there is an error, this will be an error.
+func (r *Renewer) DoneCh() <-chan error {
+	return r.doneCh
+}
+
+// RenewCh is a channel that receives a message when a successful renewal takes
+// place and includes metadata about the renewal.
+func (r *Renewer) RenewCh() <-chan *RenewOutput {
+	return r.renewCh
+}
+
+// Stop stops the renewer.
+func (r *Renewer) Stop() {
+	r.l.Lock()
+	if !r.stopped {
+		close(r.stopCh)
+		r.stopped = true
+	}
+	r.l.Unlock()
+}
+
+// Renew starts a background process for renewing this secret. When the secret
+// is has auth data, this attempts to renew the auth (token). When the secret
+// has a lease, this attempts to renew the lease.
+func (r *Renewer) Renew() {
+	var result error
+	if r.secret.Auth != nil {
+		result = r.renewAuth()
+	} else {
+		result = r.renewLease()
+	}
+
+	select {
+	case r.doneCh <- result:
+	case <-r.stopCh:
+	}
+}
+
+// renewAuth is a helper for renewing authentication.
+func (r *Renewer) renewAuth() error {
+	if !r.secret.Auth.Renewable || r.secret.Auth.ClientToken == "" {
+		return ErrRenewerNotRenewable
+	}
+
+	client, token := r.client, r.secret.Auth.ClientToken
+
+	for {
+		// Check if we are stopped.
+		select {
+		case <-r.stopCh:
+			return nil
+		default:
+		}
+
+		// Renew the auth.
+		renewal, err := client.Auth().Token().RenewTokenAsSelf(token, 0)
+		if err != nil {
+			return err
+		}
+
+		// Push a message that a renewal took place.
+		select {
+		case r.renewCh <- &RenewOutput{time.Now().UTC(), renewal}:
+		default:
+		}
+
+		// Somehow, sometimes, this happens.
+		if renewal == nil || renewal.Auth == nil {
+			return ErrRenewerNoSecretData
+		}
+
+		// Do nothing if we are not renewable
+		if !renewal.Auth.Renewable {
+			return ErrRenewerNotRenewable
+		}
+
+		// Grab the lease duration and sleep duration - note that we grab the auth
+		// lease duration, not the secret lease duration.
+		leaseDuration := time.Duration(renewal.Auth.LeaseDuration) * time.Second
+		sleepDuration := r.sleepDuration(leaseDuration)
+
+		// If we are within grace, return now.
+		if leaseDuration <= r.grace || sleepDuration <= r.grace {
+			return nil
+		}
+
+		select {
+		case <-r.stopCh:
+			return nil
+		case <-time.After(sleepDuration):
+			continue
+		}
+	}
+}
+
+// renewLease is a helper for renewing a lease.
+func (r *Renewer) renewLease() error {
```

Comment:
I felt that way too, but I cannot think of a better way to write it.

This works, but it's tricky. There's a lot of boolean logic that's well tested, but challenging to follow.

```go
// renewIt is a helper for renewing authentication.
func (r *Renewer) renewIt(renewable, auth bool, thing string) error {
    if !renewable || thing == "" {
        return ErrRenewerNotRenewable
    }

    client := r.client

    for {
        // Check if we are stopped.
        select {
        case <-r.stopCh:
            return nil
        default:
        }

        // Renew the auth.
        var renewal *api.Secret
        var err error
        if auth {
            renewal, err = client.Auth().Token().RenewTokenAsSelf(thing, 0)
            if err != nil {
                return err
            }
        } else {
            renewal, err = client.Sys().Renew(thing, 0)
            if err != nil {
                return err
            }
        }

        // Push a message that a renewal took place.
        select {
        case r.renewCh <- &RenewOutput{time.Now().UTC(), renewal}:
        default:
        }

        // Somehow, sometimes, this happens.
        if renewal == nil || (auth && renewal.Auth == nil) {
            return ErrRenewerNoSecretData
        }

        // Do nothing if we are not renewable
        if (auth && !renewal.Auth.Renewable) || !renewal.Renewable {
            return ErrRenewerNotRenewable
        }

        // Grab the lease duration and sleep duration - note that we grab the auth
        // lease duration, not the secret lease duration.
        var leaseDuration time.Duration
        if auth {
            leaseDuration = time.Duration(renewal.Auth.LeaseDuration) * time.Second
        } else {
            leaseDuration = time.Duration(renewal.LeaseDuration) * time.Second
        }
        sleepDuration := r.sleepDuration(leaseDuration)

        // If we are within grace, return now.
        if leaseDuration <= r.grace || sleepDuration <= r.grace {
            return nil
        }

        select {
        case <-r.stopCh:
            return nil
        case <-time.After(sleepDuration):
            continue
        }
    }
}
```

I tried writing something that accepted functions, but the fields you have to check for a successful auth are different from a successful renewal (which is one of the main reasons why I wrote this renewer).